[PW-6545] Error on loading undefined CC data due to invalid flow

tnaber opened the pull request:
<!-- Thank you for considering contributing to this repository! We encourage you to use PSR-2. -->

**Description**
When the payment methods have not been fully loaded yet, the renderSecureFields method can crash as the checkoutComponent has not been created yet. This PR adds a new observable for the checkoutComponent that will trigger the html to be loaded only if the checkoutComponent is defined. Thus, when the renderSecureFields (now renderCCPaymentMethod) is called the checkoutComponent will always be avaialble

This PR also introduces the adyen-checkout file which abstracts some code shared by the three 'adyen-*-method' files.

**Tested scenario's**
- [x] HPP methods
- [x] Oneclick methods
- [x] CC methods
- [x] 3DS2 modals
- [x] Multishipping
- [x] PayPal

**Fixed issue**:  <!-- #-prefixed issue number -->
Closes #1440
Closes #1445

tnaber commented on view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js:
```diff
@@ -564,64 +564,68 @@ define(
-                        houseNumber: houseNumber.join(' ').trim()
+            getFormattedAddress: function (address) {
+                function getStreetAndHouseNumberFromAddress(address, houseNumberStreetLine, customerStreetLinesEnabled) {
+                    if (typeof address.street === 'undefined') {
+                        address.street = [''];
+                    }
```
These are the actually added lines, the rest is reformatting

tnaber commented:
Hi @jonathanribas,
We're still looking into how to trigger this problem. Does you checkout page allow the payment methods to be loaded without an address being entered yet? E.g. having the payment methods and address detail fields on the same page?

jonathanribas commented:
Hi @tnaber, we actually use Adobe Commerce checkout with several steps (we don't have a One Step Checkout).

Payment methods are actually triggered by Adyen's module mixin: `vendor/adyen/module-payment/view/frontend/web/js/model/set-shipping-information-mixin.js`, as we have shipping cost estimation on our cart, we get Adyen Payment methods on cart.

If you want you can test it on https://uk.caudalie.com/

tnaber commented:
Hi @jonathanribas, 
I've found a new solution, it's not pretty so I'm leaving this as a draft for now in order to fully verify. If you are looking for a temporary solution feel free to check it out.

jonathanribas commented:
Hi @tnaber, thank you! At the moment we will keep our temporary workaround as we don't had other issues since implemented.
Looking forward the final solution from your side, maybe for next release 8.2.2?

jonathanribas commented:
Hi @tnaber, with those changes, do you still need to use `afterRender` on `cc-form.html`?

tnaber commented:
We might, as the field does not necessarily have to be available when the loadCheckoutComponent is ran. E.g. if we delete the afterRender and the loadCheckoutComponent function is ran for the last time before the field is available it will never retrigger the mount. 
However, I do agree that a single trigger would be neater and I'm still testing if we can do without one of both

kbakkes commented:
Hi, We are currently facing the same as described above.

tnaber commented:
Hi @kbakkes and @jonathanribas,

I have updated the fix and implemented a new flow. I am confident this should fix the issue.
If you would be able to test it that would be greatly appreciated!

Cheers,
Titus

michaelpaul commented on view/frontend/web/js/view/payment/method-renderer/adyen-cc-method.js:
```diff
@@ -70,50 +70,55 @@ define(
+                if (!!this.checkoutComponent) {
+                    // Setting the adyenCCMethod will load the html and trigger the renderPaymentMethod on afterRender
+                    self.adyenCCMethod({
+                            icon: !!paymentMethodsResponse.paymentMethodsExtraDetails.card.icon
+                                ? paymentMethodsResponse.paymentMethodsExtraDetails.card.icon
+                                : {}
```
```suggestion
                            icon: paymentMethodsResponse.paymentMethodsExtraDetails.card.icon || {}
```
